sdk/core/azure-core/azure/core/tracing/context.py

```diff
@@ -0,0 +1,161 @@
+import threading
+
+try:
+    from typing import TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    from typing import Any, Callable
+    from typing_extensions import Protocol
+else:
+    Protocol = object
+
+try:
+    import contextvars
+except ImportError:
+    contextvars = None
+
+
+class ContextProtocol(Protocol):
+    """
+     Implements set and get variables in a thread safe way.
+     """
+
+    def __init__(self, name, default, lock):
+        # type: (string, Any, threading.Lock) -> None
+        pass
+
+    def clear(self):
+        # type: () -> None
+        """Reset the value to the default value"""
+        pass
+
+    def get(self):
+        # type: () -> Any
+        """Get the stored value."""
+        pass
+
+    def set(self, value):
+        # type: (Any) -> None
+        """Set the value in the context."""
+        pass
+
+
+class _AsyncContext:
+    """
+    Uses contextvars to set and get variables globally in a thread safe way.
+    """
+
+    def __init__(self, name, default, lock):
+        self.name = name
+        self.contextvar = contextvars.ContextVar(name)
+        self.default = default if callable(default) else (lambda: default)
+        self.lock = lock
+
+    def clear(self):
+        # type: () -> None
+        """Reset the value to the default value"""
+        self.contextvar.set(self.default())
+
+    def get(self):
+        # type: () -> Any
+        """Get the stored value."""
+        try:
+            return self.contextvar.get()
+        except LookupError:
+            value = self.default()
+            self.set(value)
+            return value
+
+    def set(self, value):
+        # type: (Any) -> None
+        """Set the value in the context."""
+        with self.lock:
+            self.contextvar.set(value)
+
+
+class _ThreadLocalContext:
+    """
+    Uses thread local storage to set and get variables globally in a thread safe way.
+    """
+    _thread_local = threading.local()
+
+    def __init__(self, name, default, lock):
+        # type: (str, Any, threading.Lock) -> None
+        self.name = name
+        self.default = default if callable(default) else (lambda: default)
+        self.lock = lock
+
+    def clear(self):
+        # type: () -> None
+        """Reset the value to the default value"""
+        setattr(self._thread_local, self.name, self.default())
+
+    def get(self):
+        # type: () -> Any
+        """Get the stored value."""
+        try:
+            return getattr(self._thread_local, self.name)
+        except AttributeError:
+            value = self.default()
+            self.set(value)
+            return value
+
+    def set(self, value):
+        # type: (Any) -> None
+        """Set the value in the context."""
+        with self.lock:
+            setattr(self._thread_local, self.name, value)
+
+
+class TracingContext:
+    _lock = threading.Lock()
+    _context = _AsyncContext if contextvars else _ThreadLocalContext
```

Comment:
could consider `_get_context_class` function, then that function could also have the clunky contextvars import text in it as well so everything is in one place

Reply:
I like that idea! Did it!